src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java

```diff
@@ -0,0 +1,574 @@
+import java.lang.constant.DirectMethodHandleDesc;
+import java.lang.constant.DynamicCallSiteDesc;
+import java.lang.constant.DynamicConstantDesc;
+import java.lang.constant.MethodTypeDesc;
+import java.util.Collection;
+import java.util.List;
+
+import jdk.internal.classfile.BootstrapMethodEntry;
+import jdk.internal.classfile.BufWriter;
+import jdk.internal.classfile.ClassBuilder;
+import jdk.internal.classfile.ClassModel;
+import jdk.internal.classfile.ClassReader;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.impl.ClassReaderImpl;
+import jdk.internal.classfile.impl.Options;
+import jdk.internal.classfile.jdktypes.ModuleDesc;
+import jdk.internal.classfile.jdktypes.PackageDesc;
+import jdk.internal.classfile.WritableElement;
+import jdk.internal.classfile.impl.SplitConstantPool;
+import jdk.internal.classfile.impl.TemporaryConstantPool;
+import jdk.internal.classfile.impl.Util;
+
+/**
+ * Builder for the constant pool of a classfile.  Provides read and write access
+ * to the constant pool that is being built.  Writing is append-only and idempotent
+ * (entry-bearing methods will return an existing entry if there is one).
+ *
+ * A {@linkplain ConstantPoolBuilder} is associated with a {@link ClassBuilder}.
+ * The {@linkplain ConstantPoolBuilder} also provides access to some of the
+ * state of the {@linkplain ClassBuilder}, such as classfile processing options.
+ */
+public sealed interface ConstantPoolBuilder
+        extends ConstantPool, WritableElement<ConstantPool>
+        permits SplitConstantPool, TemporaryConstantPool {
+
+    /**
+     * {@return a new constant pool builder}  The new constant pool builder
+     * will inherit the classfile processing options of the specified class.
+     * If the processing options include {@link Classfile.Option.Key#CP_SHARING},
+     * (the default) the new constant pool builder will be also be pre-populated with the
+     * contents of the constant pool associated with the class reader.
+     *
+     * @param classModel the class to copy from
+     */
+    static ConstantPoolBuilder of(ClassModel classModel) {
+        ClassReader reader = (ClassReader) classModel.constantPool();
+        return reader.optionValue(Classfile.Option.Key.CP_SHARING)
+          ? new SplitConstantPool(reader)
+          : new SplitConstantPool(((ClassReaderImpl) reader).options());
+    }
+
+    /**
+     * {@return a new constant pool builder}  The new constant pool builder
+     * will be empty and have the specified classfile processing options.
+     *
+     * @param options the processing options
+     */
+    static ConstantPoolBuilder of(Collection<Classfile.Option> options) {
+        return new SplitConstantPool(new Options(options));
+    }
+
+    /**
+     * {@return the value of the specified option}
+     *
+     * @param option the key of the option value
+     * @param <T> the type of the option value
+     */
+    <T> T optionValue(Classfile.Option.Key option);
```

Comment:
This unconstrained type parameter will result in and implicit conversion to any type that the caller assign it to, which might result in a `ClassCastException` if the caller gets the type wrong.

Is that intentional?

An alternative solutions: 

* Convert `Key` to an ordinary class and add a type parameter to it, for the value type.
* Add a parameter of type `Class<T>`.

Reply:
Thanks for pointing it out.
I've re-visited the API re-accessing already provided options and found no relevant use case.
This method (and similar on other places), plus the `Classfile.Option.Key` will be pulled from API to the implementation.
Please let us know if there is a need to read option values from user perspective.